### PR TITLE
pluginmanager.consider_preparse: add exclude_only kwarg

### PR DESCRIPTION
**Pull request by blueyed**

### CODE REVIEW

**Comment by blueyed** [comment]
> this moves the consideration of `-p` plugins from before to after plugins

Do you mean the internal plugins?
(entrypoint plugins clearly get loaded afterwards)

I've quickly checked with `foo.py` and `pytest -p foo`: import order is not affected it seems, but `pytest_configure` gets called _before_ e.g. the terminal plugin with this branch, but _after_ the terminal plugin on master.

**Comment by RonnyPfannschmidt** [comment]
@blueyed that's what i meant - this makes a lot of things more sane - but https://xkcd.com/1172/ comes to mind

**Comment by blueyed** [comment]
@RonnyPfannschmidt I see.  I was confused since I've observed another order than you said.
I cannot really tell what is more sane, but noticed this with the `load_entrypoint_plugins` marker (https://github.com/blueyed/pytest/pull/71), IIRC.

**Comment by nicoddemus** [comment]
Hey guys,

Sorry, I'm not sure I follow. Adding a `foo.py` file with `pytest_configure()` and executing `pytest -p foo` I always see `foo.pytest_configure` being called before `terminal.pytest_configure`. 

(Btw, @blueyed please add some description when opening PRs, sometimes it is hard to divine the intent of a PR without some context; one or two phrases can save everyone some time, both reviewers trying to figure out what the PR is supposed to be about, and the ensuing back and forth that might happen afterwards in case there's questions about the purpose of the PR).

**Comment by blueyed** [comment]
@nicoddemus see https://github.com/pytest-dev/pytest/pull/6534 for a test in this regard - it changes with this PR here (https://github.com/blueyed/pytest/commit/c5f4d6b108361cf73c44188c4f6f033f71f58310).

**Comment by blueyed** [comment]
@nicoddemus this PR changes it to before, without it it comes after (IIRC) - I've elaborated on that before (makes you asking for more descriptive things a bit funny, but will keep it in mind - although I think I am being verbose in general).

**Comment by nicoddemus** [comment]
> @nicoddemus this PR changes it to before, without it it comes after

I see, thanks, I was seeing something different it seems, probably messed something on my side. Now I'm seeing what you describe, on this branch:

```
λ python -m pytest test-ro.py -p foo -q
foo pytest_configure
terminal pytest_configure
.
1 passed in 0.01s
```

`foo` is registered after internal plugins, so its `pytest_configure` gets called first (hooks are called in inverse order of registration).

On `features`:

```
λ python -m pytest test-ro.py -p foo -q
terminal pytest_configure
foo pytest_configure
.
1 passed in 0.01s
```

But the `-p` option explicitly says:

```
  -p name               early-load given plugin module name or entry point (multi-allowed). To avoid
                        loading of plugins, use the `no:` prefix, e.g. `no:doctest`.
```

So it seems the behavior is intentional, no?

**Comment by blueyed** [comment]
> early-load given plugin module name

It does not say though "how early".  But I can see that this better describes the (new) behavior here, and that's likely why @RonnyPfannschmidt considers it to be more sane.

**Comment by nicoddemus** [comment]
> It does not say though "how early".

I agree, it is open to interpretation.

I think it is OK to merge this and release it in `5.4`; I suspect *very* few cases should be affected (if at all), and then we can decide on how to proceed then.

**Comment by blueyed** [comment]
Added a changelog ("breaking"), adjusted `test_plugin_loading_order`, and rebased.

File: changelog/6443.breaking.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Plugins specified with ``-p`` are now loaded before internal ones.
```
Comment (nicoddemus) [inline changelog/6443.breaking.rst]:
```suggestion
Plugins specified with ``-p`` are now loaded after internal plugins, meaning their hooks will be called *before* the internal ones. 

This makes the ``-p`` behavior consistent with ``PYTEST_PLUGINS``.
```

**Pull request merged**